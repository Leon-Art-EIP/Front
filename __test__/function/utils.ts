import type { ReactTestRendererJSON } from "react-test-renderer";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import { expect } from "vitest";

function toJson(component: renderer.ReactTestRenderer): ReactTestRendererJSON {
  const result = component.toJSON();
  expect(result).toBeDefined();

  return result as ReactTestRendererJSON;
}

export function testSnapshot(element: JSX.Element): void {
  const component = renderer.create(element);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
}
