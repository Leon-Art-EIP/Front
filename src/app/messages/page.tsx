export default function Page(): JSX.Element {
  return <div>Page de messages</div>;
}
