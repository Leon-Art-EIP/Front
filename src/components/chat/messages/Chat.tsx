export interface ChatProps {
  content: string;
  sender: 0 | 1;
  dateTime: string;
}

export function Chat(props: ChatProps): JSX.Element {
  /* c8 ignore start */
  return (
    <div className={`flex flex-col ${props.sender ? "items-start" : "items-end"} justify-center px-6`}>
      <div
        className={`rounded-2xl p-4 max-w-xl break-words ${
          props.sender ? "bg-background-hl text-tertiary" : "bg-primary text-white"
        }`}
      >
        {props.content}
      </div>
      {props.dateTime && <span className="text-sm px-2 py-1 text-tertiary">{props.dateTime}</span>}
    </div>
  );
}
/* c8 ignore stop */
