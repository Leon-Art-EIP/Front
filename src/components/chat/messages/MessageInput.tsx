import SendIcon from "@mui/icons-material/Send";
import { useState } from "react";

export interface MessageInputProps {
  handleSendMsg: (msg: string) => void;
}
export function MessageInput(props: MessageInputProps): JSX.Element {
  const [messageToSend, setMessageToSend] = useState("");

  function onSendMessage(e: React.FormEvent) {
    e.preventDefault();

    if (messageToSend === "") return;
    props.handleSendMsg(messageToSend);
    setMessageToSend("");
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLTextAreaElement>) {
    if (e.key === "Enter" && !e.shiftKey) {
      e.preventDefault();
      onSendMessage(e as unknown as React.FormEvent);
    }
  }

  return (
    <form onSubmit={onSendMessage} className="flex justify-center items-end w-full gap-4 px-6 pb-6 pt-2 xl:px-10 xl:pb-6 xl:pt-2">
      <textarea
        value={messageToSend}
        onKeyDown={handleKeyDown}
        onChange={(e) => setMessageToSend(e.target.value)}
        placeholder="Votre message..."
        className="text-tertiary w-full bg-background-inputfield outline-none placeholder:text-tertiary placeholder:opacity-50 resize-y overflow-y-auto rounded-3xl px-8 py-4"
        style={{ height: "3.5rem", minHeight: "3.5rem", maxHeight: "14rem" }}
      />
      <button type="submit" className="bg-quaternary-hover p-3 rounded-full text-center">
        <SendIcon className="h-7 w-7 fill-quaternary" />
      </button>
    </form>
  );
}