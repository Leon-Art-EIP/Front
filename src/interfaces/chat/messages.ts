export interface IMessage {
  id: number;
  senderId: string;
  contentType: "text" | "image";
  content: string;
  dateTime: string;
  read: boolean;
}

export interface IOrderInfos {
  userRole: string;
  orderState: string;
  orderRating: number;
  orderPicture: string;
  orderDescription: string;
  orderPrice: number;
}