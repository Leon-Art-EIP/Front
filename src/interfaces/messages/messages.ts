export interface IMessages {
  messages: IMessage[];
}

export interface IMessage {
  id: number;
  sender: number;
  contentType: string;
  content: string;
  date: string;
  read: boolean;
}