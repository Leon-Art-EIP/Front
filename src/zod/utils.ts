export const getValueOrUndefined = (arg: unknown) =>
  arg === "" || arg === undefined || arg === null ? undefined : arg;
