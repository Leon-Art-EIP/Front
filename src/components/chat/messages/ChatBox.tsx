import { useEffect, useRef } from "react";
import { useChat } from "../../../contexts/ChatContext";
import { Chat } from "./Chat";

export function ChatBox(): JSX.Element {
  /* c8 ignore start */
  const { messages, currentUser } = useChat() || {};

  const chatBoxRef = useRef<HTMLDivElement>(null);
  let prevDateTime: Date | null = null;

  useEffect(() => {
    if (chatBoxRef.current) {
      chatBoxRef.current.scrollTop = chatBoxRef.current.scrollHeight;
    }
  }, [messages]);

  const formatDate = (string: string) => {
    const options: Intl.DateTimeFormatOptions = { year: "numeric", month: "long", day: "numeric" };
    return new Date(string).toLocaleDateString([], options);
  };

  const formatTime = (string: string) => {
    return new Date(string).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" });
  };

  return (
    <div ref={chatBoxRef} dir="btt" className="flex flex-col h-full w-full overflow-y-auto gap-4">
      {messages.map((message, index) => {
        const currentDateTime = new Date(message.dateTime);
        let showFullDate = false;

        if (prevDateTime) {
          if (currentDateTime.toDateString() !== prevDateTime.toDateString()) {
            showFullDate = true;
          }
        } else {
          showFullDate = true;
        }

        prevDateTime = currentDateTime;

        return (
          <div key={index} className="flex flex-col">
            {showFullDate && (
              <>
                <span className="self-center mt-3 h-1 w-1/3 bg-background-hl rounded-full" />
                <span className="text-lg text-center p-2 text-background-hl">{formatDate(message.dateTime)}</span>
              </>
            )}
            <Chat
              dateTime={formatTime(message.dateTime)}
              key={message.id}
              content={message.content}
              sender={message.senderId === currentUser?.user.id ? 0 : 1}
            />
          </div>
        );
      })}
    </div>
  );
}
/* c8 ignore stop */
