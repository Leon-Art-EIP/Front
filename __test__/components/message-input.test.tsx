import { test } from "vitest";
import { testSnapshot } from "../function/utils";
import { MessageInput } from "../../src/components/chat/messages/MessageInput";

test("ChatBox snapshot", () => {
  testSnapshot(<MessageInput handleSendMsg={() => {}} />);
});
