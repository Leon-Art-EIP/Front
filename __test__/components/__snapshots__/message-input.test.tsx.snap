// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatBox snapshot 1`] = `
<form
  className="flex justify-center items-end w-full gap-4 px-6 pb-6 pt-2 xl:px-10 xl:pb-6 xl:pt-2"
  onSubmit={[Function]}
>
  <textarea
    className="text-tertiary w-full bg-background-inputfield outline-none placeholder:text-tertiary placeholder:opacity-50 resize-y overflow-y-auto rounded-3xl px-8 py-4"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Votre message..."
    style={
      {
        "height": "3.5rem",
        "maxHeight": "14rem",
        "minHeight": "3.5rem",
      }
    }
    value=""
  />
  <button
    className="bg-quaternary-hover p-3 rounded-full text-center"
    type="submit"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium h-7 w-7 fill-quaternary css-i4bv87-MuiSvgIcon-root"
      data-testid="SendIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
      />
    </svg>
  </button>
</form>
`;
